src/wallet/wallet.cpp

```diff
@@ -857,80 +858,128 @@ bool CWallet::LoadUnifiedAddressMetadata(const ZcashdUnifiedAddressMetadata &add
+            nHeight = pTxIndex->nHeight;
+        }
+    }
+
+    auto ufvk = self->GetUFVKForReceiver(RecipientAddressToReceiver(recipient));
+    std::pair<PaymentAddress, RecipientType> defaultAddress = std::visit(match {
+        [&](const CKeyID& addr) {
+            auto ua = self->FindUnifiedAddressByReceiver(addr);
+            if (ua.has_value()) {
+                // we do not generate unified addresses for change keys
+                return std::make_pair(PaymentAddress{ua.value()}, RecipientType::WalletExternalAddress);
+            }
+
+            // not associated with an UA, so we check legacy key metadata; if we don't
+            // have any metadata, it's a counterparty address
+            auto keyMeta = mapKeyMetadata.find(addr);
+            if (keyMeta == mapKeyMetadata.end()) {
+                return std::make_pair(PaymentAddress{addr}, RecipientType::CounterpartyAddress);
+            }
+
+            // for legacy change keys, we can check the keypath to determine
+            // whether this is an internal or external address.
+            if (IsInternalKeyPath(44, BIP44CoinType(), keyMeta->second.hdKeypath)) {
+                return std::make_pair(PaymentAddress{addr}, RecipientType::WalletInternalAddress);
+            } else {
+                return std::make_pair(PaymentAddress{addr}, RecipientType::WalletExternalAddress);
+            }
+        },
+        [&](const CScriptID& addr) {
+            auto ua = self->FindUnifiedAddressByReceiver(addr);
+            if (ua.has_value()) {
+                return std::make_pair(PaymentAddress{ua.value()}, RecipientType::WalletExternalAddress);
+            } else if (HaveCScript(addr)) {
+                // we never generate internal p2sh addresses, so all are wallet external
+                return std::make_pair(PaymentAddress{addr}, RecipientType::WalletExternalAddress);
+            }
+
+            return std::make_pair(PaymentAddress{addr}, RecipientType::CounterpartyAddress);
+        },
+        [&](const SaplingPaymentAddress& addr) {
+            auto ua = self->FindUnifiedAddressByReceiver(addr);
+            if (ua.has_value()) {
+                // UAs are always external addresses
+                return std::make_pair(PaymentAddress{ua.value()}, RecipientType::WalletExternalAddress);
+            } else if (ufvk.has_value() && ufvk->GetSaplingKey().has_value()) {
+                auto saplingKey = ufvk->GetSaplingKey().value();
+                auto j = saplingKey.DecryptDiversifier(addr);
+                if (j.has_value()) {
+                    // external addresses correspond to UAs.
+                    if (j.value().second) {
+                        // std::get is safe here because we know we have a valid Sapling diversifier index
+                        auto defaultUA = std::get<std::pair<UnifiedAddress, diversifier_index_t>>(
+                                ufvk->Address(j.value().first, CWallet::DefaultReceiverTypes(nHeight)));
```

Comment:
Technically we should choose the default receiver types making sure they include the receiver we know is there. But our current set of receiver types encompasses everything we generate, so this is currently fine. Open an issue to fix this.

Reply:
https://github.com/zcash/zcash/issues/5722